operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeploymentDependentResource.java

```diff
@@ -287,6 +287,7 @@ private StatefulSet createBaseDeployment(Keycloak keycloakCR, Context<Keycloak>
                         .editOrNewSpec().withImagePullSecrets(keycloakCR.getSpec().getImagePullSecrets()).endSpec()
                     .endTemplate()
                     .withReplicas(keycloakCR.getSpec().getInstances())
+                    .withServiceName(KeycloakDiscoveryServiceDependentResource.getName(keycloakCR))
```

Comment:
Should it rather use the main service?
```suggestion
                    .withServiceName(KeycloakServiceDependentResource.getServiceName(keycloakCR))
```

Reply:
No, it's supposed to be a headless service.

Reply:
From the [docs|https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/stateful-set-v1/#StatefulSetSpec] it does not seem so?

Additionally it says:
> This service must exist before the StatefulSet

Despite it now works with StS created before the Service, can we really rely on it as that behaviour might change to align with the spec?

Reply:
> From the [docs](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/stateful-set-v1/#StatefulSetSpec) it does not seem so?

I'm referencing https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id 

I suppose the language there does say "can", not "must" be headless.

> Despite it now works with StS created before the Service, can we really rely on it as that behaviour might change to align with the spec?

This behavior / validation has been poorly defined for quite some time. https://github.com/kubernetes/kubernetes/issues/69608 was logged years ago.

Even if it doesn't "work" at some point in the furture, the worst case behavior would be the statefulset not working until scaling up until the service is created - since these are peer dependent resources, that would be a narrow window. 

I'm leary of adding any kind of reconciliation condition for this kind of behavior - instead if you are worried about not eventually seeing some kind of ephemeral error, then I'd opt instead for Dependent.dependsOn

Reply:
> Even if it doesn't "work" at some point in the furture, the worst case behavior would be the statefulset not scaling up until the service is created - since these are peer dependent resources, that would be a narrow window.

Makes sense 👍

Reply:
> We might create a dedicated headlessService just for this purpose but that seems like an overkill at this point.

I think we'd just leave the discovery service in place. There's not much harm / risk and users may have already adopted using it for their own purposes as well.